pkg/services/alertmanager.go

```diff
@@ -0,0 +1,208 @@
+				return err
+			}
+		}
+		if len(n.Annotations) > 0 {
+			notification.Alertmanager.Annotations = n.Annotations
+			if err := notification.Alertmanager.parseAnnotations(name, f, vars); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}, nil
+}
+
+func (n *AlertmanagerNotification) parseAnnotations(name string, f texttemplate.FuncMap, vars map[string]interface{}) error {
+	for k, v := range n.Annotations {
+		var tempData bytes.Buffer
+		tmpl, err := texttemplate.New(name).Funcs(f).Parse(v)
+		if err != nil {
+			return err
+		}
+		if err := tmpl.Execute(&tempData, vars); err != nil {
+			return err
+		}
+		if val := tempData.String(); val != "" {
+			n.Annotations[k] = val
+		}
+	}
+
+	return nil
+}
+
+func (n *AlertmanagerNotification) parseLabels(name string, f texttemplate.FuncMap, vars map[string]interface{}) error {
+	foundAlertname := false
+
+	for k, v := range n.Labels {
+		if k == alertNameLabel {
+			foundAlertname = true
+		}
+
+		var tempData bytes.Buffer
+		tmpl, err := texttemplate.New(name).Funcs(f).Parse(v)
+		if err != nil {
+			return err
+		}
+		if err := tmpl.Execute(&tempData, vars); err != nil {
+			return err
+		}
+		if val := tempData.String(); val != "" {
+			n.Labels[k] = val
+		}
+	}
+
+	if !foundAlertname {
+		n.Labels[alertNameLabel] = name
+	}
+
+	return nil
+}
+
+// Send using create alertmanager events
+func (s alertmanagerService) Send(notification Notification, dest Destination) error {
+	rawBody, err := json.Marshal([]*AlertmanagerNotification{notification.Alertmanager})
+	if err != nil {
+		return err
+	}
+
+	for _, target := range s.opts.Targets {
+		go func(target string) {
+			if err := s.sendOneTarget(target, rawBody); err != nil {
+				s.entry.Errorf("alertmanager sent target: %v", err)
+			}
+		}(target)
+	}
+
+	return nil
+}
+
+func (s alertmanagerService) sendOneTarget(target string, rawBody []byte) error {
+	rawURL := fmt.Sprintf("%v://%v%v", s.opts.Scheme, target, s.opts.APIPath)
+
+	transport := httputil.NewTransport(rawURL, s.opts.InsecureSkipVerify)
+	client := &http.Client{
+		Transport: httputil.NewLoggingRoundTripper(transport, s.entry),
+	}
+
+	req, err := http.NewRequest(http.MethodPost, rawURL, bytes.NewReader(rawBody))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	if s.opts.BasicAuth != nil {
+		req.SetBasicAuth(s.opts.BasicAuth.Username, s.opts.BasicAuth.Password)
+	}
+	if s.opts.BearerToken != "" {
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %v", s.opts.BearerToken))
+	}
+
+	response, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		data, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			data = []byte(fmt.Sprintf("unable to read response data: %v", err))
+		}
+		return fmt.Errorf("request to %s has failed with error code %d : %s", rawURL, response.StatusCode, string(data))
+	}
+
+	defer func() {
```

Comment:
You need to close response body on line 194 because this code hasn’t close body when the error happened.
Also if you don't use body you need to drop body as follows.

```
defer func() {
  io.Copy(ioutil.Discard, resp.Body)
  resp.Body.Close()
}()
```